### PR TITLE
Proposal for the new Kitchen Sink "Textarea" component

### PR DESCRIPTION
**Pull request by xus**
Rules
https://github.com/ILIAS-eLearning/ILIAS/pull/1028/files

### CODE REVIEW

File: src/UI/Implementation/Component/Input/Field/TextArea.php
Comment (klees) [inline src/UI/Implementation/Component/Input/Field/TextArea.php]: Is a length of 1 always enough to fulfill a required textarea? Maybe the component needs a parameter for that purpose, but I'm not really sure.

File: src/UI/examples/Input/Field/TextArea/base.php
Comment (klees) [inline src/UI/examples/Input/Field/TextArea/base.php]: Don't mix indentation, please.

**Review by klees — CHANGES_REQUESTED** [review]
Hi @xus , thx again for the contribution. This LGTM in general added one minor food for thought and a minor formatting problem only. Please solve the latter. I really like that the implementation looks that simple. Thx!

File: src/UI/Implementation/Component/Input/Field/TextArea.php
Comment (xus) [reply src/UI/Implementation/Component/Input/Field/TextArea.php]: Yes, I think you are right. Textareas can be limited also with a maximum number of characters. I will create the constraint "hasMaxLenght" and deal with the proper parameters to set the values. 

Edit-> Now we have this new constraint so we can limit the textareas with min and max number of characters as the current implementation.

**Comment by xus** [comment]
Todo:
- [ ] PHPUnit tests
- [x] Error messages using lang vars. (Max and Min length limits)
- [x] JS counter showing the number of characters available when writing.
- [ ] Use language variables e.g. Constraints\HasMaxLenght.php

File: src/UI/Component/Input/Field/Factory.php
Comment (Amstutz) [inline src/UI/Component/Input/Field/Factory.php]: Please add something like: If a limit is set, this limit MUST be indicate this somehow (e.g. by stating this in the byline).

File: src/UI/Implementation/Component/Input/Field/TextArea.php
Comment (Amstutz) [reply src/UI/Implementation/Component/Input/Field/TextArea.php]: nice.

File: src/UI/examples/Input/Field/TextArea/with_limits.php
Comment (Amstutz) [inline src/UI/examples/Input/Field/TextArea/with_limits.php]: Please use another name than example_name here, due to the first example. So the display of the examples will run a little smoother.

**Review by Amstutz — CHANGES_REQUESTED** [review]
Thanks for this contribution. Good examples! 

I have one little note rerarding the indication of max length to the user. I think at least in the comment this should be stated if used.

File: src/UI/Component/Input/Field/Factory.php
Comment (xus) [reply src/UI/Component/Input/Field/Factory.php]: Thanks for this comment, I will change this line. The byline will be set depending if the component has max/min limit, with the following texts:
e.g.
Minimum number of characters: 10 Maximum number of characters: 100

File: src/UI/Component/Input/Field/Factory.php
Comment (xus) [inline src/UI/Component/Input/Field/Factory.php:216]: Should we mention the byline and the line showing how many characters are used ( if limit is set) ?

File: src/UI/Component/Input/Field/Factory.php
Comment (matthiaskunkel) [inline src/UI/Component/Input/Field/Factory.php]: Jour Fixe, 23 APR 2018 : Please add to rules: "If a min. or max. number of characters is set for textarea, a byline MUST be added stating the number of min. and/or max. characters."

File: src/UI/Component/Input/Field/Factory.php
Comment (matthiaskunkel) [inline src/UI/Component/Input/Field/Factory.php:216]: "If a limit is set, a byline about limitation is automatically set."

**Comment by matthiaskunkel** [comment]
Jour Fixe, 23 APR 2018 : PR is highly appreciated. Please consider the requirements of the Jour Fixe before committing this PR to trunk.

File: src/UI/Component/Input/Field/Factory.php
Comment (matthiaskunkel) [inline src/UI/Component/Input/Field/Factory.php]: JF "A textarea counts the amount of character input by user and displays the number."

**Comment by matthiaskunkel** [comment]
Jour Fixe, 23 APR 2018 : PR is highly appreciated. Please consider the requirements of the Jour Fixe before committing this PR to trunk.

**Comment by Amstutz** [comment]
Is this ready for a final review?

**Comment by xus** [comment]
@Amstutz Not yet, more work should be done.

File: src/UI/Implementation/Component/Input/Field/Renderer.php
Comment (xus) [inline src/UI/Implementation/Component/Input/Field/Renderer.php]: I have to deal with language variables in different places and some of them should be used in the JS. IMO this is something that we will need in other scenarios. @Amstutz  @klees Are you OK with loading the language module in the Form container?
/src/UI/Implementation/Component/Input/Container/Form/Form.php

public function __construct(array $inputs) {
        global $DIC;
        $this->lng = $DIC->language();
        $this->lng->loadLanguageModule("form");
        ...
    ...

Comment (Amstutz) [reply src/UI/Implementation/Component/Input/Field/Renderer.php]: Hi @xus. you should not use the $DIC in the renderer on the field. The DI here works hierarchically, meaning we inject the dependencies directly on the abstractComponentRenderer. Meaning you have access to the correct language dependency through $this->lng (inside the renderer). The component should not be aware of any language. 

You have a valid point with the language module. Obviously, there is some central place, needing to load the language module for the form. The renderer of the form would probably be the right place to do that. Maybe with a nice if before to check if already loaded. @klees do you agree?

File: src/Validation/Constraints/HasMaxLength.php
Comment (xus) [inline src/Validation/Constraints/HasMaxLength.php:22]: Hi @Amstutz, @klees  All the UI Constraints are using hard coded messages(english), nobody is using lang vars. Is multi-language available somehow? is this a bug or it's something that are you working on?

File: src/UI/Implementation/Component/Input/Field/Renderer.php
Comment (klees) [reply src/UI/Implementation/Component/Input/Field/Renderer.php]: In general I'm fine with @Amstutz's solution. But I still would like to understand the general problem a bit more. @xus: What are these lang vars you want to use? How are they form specific and not e.g. part of "common"?

File: src/Validation/Constraints/HasMaxLength.php
Comment (klees) [reply src/Validation/Constraints/HasMaxLength.php:22]: This is an issue I'm working on. I already have some code ready, but need [this](https://github.com/ILIAS-eLearning/ILIAS/pull/1054) to make it work. So no need for you to i18n that by yourself.

File: src/UI/Implementation/Component/Input/Field/Renderer.php
Comment (klees) [reply src/UI/Implementation/Component/Input/Field/Renderer.php]: @Amstutz AFAIK ilLanguage already checks itself if a lang module is already loaded.

Comment (xus) [reply src/UI/Implementation/Component/Input/Field/Renderer.php]: Thank you guys for your attention on this topic.
The UI textarea needs multilanguage in two places when number of characters is limited by the user:
- Using the byline to show the "Minimum:" and "Maximum" number of characters allowed.
- Displaying the character counter via JS. (in trunk -> form#:#form_chars_remaining#:#Characters remaining:)
	$lng->toJS("form_chars_remaining");

I thought that since we are using forms we should use the form module. Probably in the future other Inputs will need multilanguage and identify somehow all this lang vars can be useful.
Obviously I will be completely fine with your final decision/suggestions.

Comment (klees) [reply src/UI/Implementation/Component/Input/Field/Renderer.php]: Thx for the info @xus. I think we might need to have a look into the organization of the lang vars (for UI framework at least?) at some point in the future, but atm I'm fine with loading the "form"-module in the according renderer, as @Amstutz suggests.

Comment (xus) [reply src/UI/Implementation/Component/Input/Field/Renderer.php]: Yes, at least for the UI Framework well be nice. Thank you @klees @Amstutz

**Comment by Amstutz** [comment]
@xus Do you wait for some more feedback from our side? What is that status, are you still working on it?

**Comment by xus** [comment]
Hi @Amstutz, I''m waiting for feedback about what I mentioned in the previous comment about how to deal with the lang vars and if the PR mentioned by @klees will provide a solution for this issue.

**Comment by Amstutz** [comment]
@xus thanks for the update, other than that, this is ready for a final review?

File: src/UI/Implementation/Component/Input/Field/Renderer.php
Comment (xus) [reply src/UI/Implementation/Component/Input/Field/Renderer.php]: Hi @Amstutz, if I understood your first comment properly, you said that we can't use $DIC in the input renderer. I see that we can use directly $this->lng but I have to load a language module somehow and I don't see the way you want us to do this.(I have only access to "common" vars).

My first idea was add a new method in the AbstractComponentRenderer to do this task, something like:
final public function loadLangModule($lang)
{
	$this->lng->loadLanguageModule($lang);
}

allowing to load the module from the renderer:
protected function renderTextareaField(Template $tpl, TextArea $input)
{
	$this->loadLangModule('form');
	...
	...

But looking other Pull Requests I see that other developers are using DIC inside the specific renderers. Look at the file:
src/UI/Implementation/Component/Input/Field/Renderer.php

in PRS:
https://github.com/ILIAS-eLearning/ILIAS/pull/1142/files
https://github.com/ILIAS-eLearning/ILIAS/pull/1136/files

Can I keep the code as it is?
protected function renderTextareaField(Template $tpl, TextArea $input)
{
	global $DIC;
	$lng = $DIC->language();
	$lng->loadLanguageModule("form");
...

Comment (Amstutz) [reply src/UI/Implementation/Component/Input/Field/Renderer.php]: Hi @xus, thanks for pointing this out. I did not perform a review on the first PR yet, so I can not yet say anything about that. I will look into that when I perform the review. For the second, I did not notice. I now commented on it, thx.

Do you understand why it is important for us, that $DIC is not used through global access? Do you agree that this is important?

You are right in pointing out, that there is currently now way to load the module, since the lng in AbstractComponentRenderer is kept private. I would be fine with your suggestion of opening up the interface there slightly by allowing:
```
final public function loadLangModule($lang)
{
  $this->lng->loadLanguageModule($lang);
}
```

Comment (klees) [reply src/UI/Implementation/Component/Input/Field/Renderer.php]: I would propose to just introduce a "ui"-module and load it anyway. This also would also organize the variables nicely. Opening that interface IMO is to much. Maybe @matthiaskunkel has an opinion about an "ui"-module?

Comment (Amstutz) [reply src/UI/Implementation/Component/Input/Field/Renderer.php]: I am also fine with the ui-module, as long as @xus has some possible option to finish this task ;-).

Comment (xus) [reply src/UI/Implementation/Component/Input/Field/Renderer.php]: Thank you guys for the feedback.

> Do you understand why it is important for us, that $DIC is not used through global access? Do you agree that this is important?

Yes, I think so. I agree with you guys that injecting dependencies and avoid global vars is the way to go for multiple and safer reasons.

I will add the function in to AbstractComponentRenderer, thanks @Amstutz 

I also think that ui-module can be useful, thanks @klees

Comment (klees) [reply src/UI/Implementation/Component/Input/Field/Renderer.php]: *I will add the function in to AbstractComponentRenderer, thanks @Amstutz*

Please don't. I don't think people should be able or required to load lang modules while rendering.

Just put your lang-var in an "ui"-module and load it [here](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/src/UI/Implementation/Render/DefaultRendererFactory.php#L44).

Thx for your patience.

Comment (xus) [reply src/UI/Implementation/Component/Input/Field/Renderer.php]: Bravo!
Creating this new ui module and loading it on getRendererInContext solved all the language problems.
Thanks a lot!

**Comment by xus** [comment]
Hi @klees @Amstutz this PR is ready for your revision. Thank you so much for the provided feedback

**Review by Amstutz — CHANGES_REQUESTED** [review]
Hi @xus. I just checked out the implementation and the examples. Looking good, good work. 

Before merging:
- [ ] Please resolve remaining merge conflicts

- [ ] Add the missing unit tests. We are aware that some Input Fields are also lacking them (e.g. Checkboxes) this is bad an already noted.

**Comment by xus** [comment]
New PR without conflicts -> #1161

**Pull request closed without merging**
